question-servers/freeform.js

```diff
@@ -6,10 +6,74 @@ var path = require('path');
 var mustache = require('mustache');
 var cheerio = require('cheerio');
 
-var elements = require('./elements');
+var {
+    elements,
+    dependencies: elementDependencies
+} = require('./elements');
 var codeCaller = require('../lib/code-caller');
 
 module.exports = {
+    /**
+     * Computes the dependencies given a Set of element names.
+     * @param  {Set}      elements
+     * @param  {Function} callback Will be called with an array of <link>/<script> tags
+     */
+    getElementDependencies: function(elements, callback) {
+        const elementStyleFiles = [];
+        const elementScriptFiles = [];
+        for (const elem of elements) {
+            elementStyleFiles.push(`${elem}.css`);
+            elementScriptFiles.push(`${elem}.js`);
+        }
+        const styleUrls = new Set();
+        const scriptUrls = new Set();
+        async.parallel([
+            (callback) => {
+                async.each(elementStyleFiles, (file, callback) => {
+                    fs.access(path.join(__dirname, 'elements', file), (err) => {
+                        if (!err) {
+                            styleUrls.add(`/pl/static/elements/${file}`);
+                            //headerHtmls.push(`<link href="/pl/static/elements/${file}" rel="stylesheet" />`);
+                        }
+                        callback(null);
+                    });
+                }, () => callback(null));
+            },
+            (callback) => {
+                async.each(elementScriptFiles, (file, callback) => {
+                    fs.access(path.join(__dirname, 'elements', file), (err) => {
+                        if (!err) {
+                            scriptUrls.add(`/pl/static/elements/${file}`);
+                            //headerHtmls.push(`<script type="text/javascript" src="/pl/static/elements/${file}"></script>`);
+                        }
+                        callback(null);
+                    });
+                }, () => callback(null));
+            }
+        ], () => {
+            // Now that we've ~automagically~ added all available files, let's
+            // add any other declared dependencies
+            const globalScriptUrls = new Set();
+            for (const elem of elements) {
+                const deps = elementDependencies[elem];
+                if (!deps) continue;
+                deps.styles && deps.styles.forEach((file) => styleUrls.add(`/pl/static/elements/${file}`));
```

Comment:
What happens to dependency ordering here? I'm thinking about the case where an element says that it needs lodash and backbone, and it's essential that lodash appears before backbone in the page head because backbone uses lodash. It could also be the case that some other element just requires lodash, so lodash is added to the Set more than once.

Ideally, we should be able to guarantee the property that if all elements list dependency A before dependency B, or they only include just one of A or B, then in the final order A will be listed before B.